### PR TITLE
Convert PenaltyKickEnemyPlay to using FSMs

### LINKED ISSUE #2714: Re enable tests post robo cup
### Description of the task

Some changes broke tests
seems mostly HRVO related

### Acceptance criteria

- [ ]

### LINKED ISSUE #2916: Goalie stuck in MoveToGoalLine state during penalty kick
### Description of the task

The goalie remains in MoveToGoalLine when the ball is put into play during a penalty kick

https://github.com/UBC-Thunderbots/Software/assets/60044853/ce490136-01c4-4e98-8f28-5ca0d51048e9

### Acceptance criteria

- [ ] Fix bug and verify that the goalie transitions out of MoveToGoalLine when the ball is put into play

### LINKED ISSUE #2920: Convert PenaltyKickEnemyPlay into using FSMs
### Description of the task

PenaltyKickEnemyPlay is currently written with coroutines and we want to move all of our plays into using FSMs.

### Acceptance criteria

- [ ]

### PR DESCRIPTION
**Pull request by williamckha**
### Description

Convert PenaltyKickEnemyPlay to using FSMs and fixes issue with goalie not moving during enemy penalty kicks

### Testing Done

Tests in `penalty_kick_enemy_play_test.cpp` pass and added `penalty_kick_enemy_play_fsm_test.cpp`

### Resolved Issues

Resolves #2920 
Resolves #2916 
Resolves #2714

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_test.cpp
Comment (nimazareian) [inline src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_test.cpp:1]: I tried running the disabled test (`test_penalty_kick_enemy_play_setup`) and in the first test case all friendly robots move to their positions, but in the rest of the test cases all friendly robots are stationary but the goalie. Do you know why that is?

File: src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_fsm.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        move_tactics = std::vector<std::shared_ptr<MoveTactic>>(num_tactics);
+        std::generate(move_tactics.begin(), move_tactics.end(),
+                      []() { return std::make_shared<MoveTactic>(); });
+    }
+
+    // Move all robots behind the penalty mark
+    for (unsigned int i = 0; i < move_tactics.size(); i++)
+    {
+        double y_offset = 4 * ((double)i - ((double)move_tactics.size() - 1) / 2.0) *
```
Comment (nimazareian) [inline src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_fsm.cpp]: nit: can we increase the offset just to give the robots a bit more space to motion plan around each other?
```suggestion
        double y_offset = 8 * ((double)i - ((double)move_tactics.size() - 1) / 2.0) *
```

File: src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_test.cpp
Comment (nimazareian) [reply src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_test.cpp:1]: Might be able to enable the tests as well if we slightly modify the over-constrained validations

**Review by nimazareian — COMMENTED** [review]
Overall looks good. Just a couple of nits

File: src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_fsm.cpp (hunk 1 above)
Comment (williamckha) [reply src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_fsm.cpp]: done, also adjusted offsets in `penalty_kick_play_fsm.cpp`

File: src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_test.cpp
Comment (williamckha) [reply src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_test.cpp:1]: Looks like the referee commands weren't correct, should be fixed now
Also `behind_ball_region` being used for validation didn't look right. Fixed that and tests are reenabled + all pass now

File: src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_fsm.cpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    tactics_to_run[0].insert(tactics_to_run[0].end(), move_tactics.begin(),
+                             move_tactics.end());
+    event.common.set_tactics(tactics_to_run);
+}
+
+void PenaltyKickEnemyPlayFSM::defendKick(const Update &event)
+{
+    // Allow goalie to move freely and defend against penalty kick
+    event.control_params.goalie_tactic->updateControlParams(false);
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_fsm.cpp:42]: nice!

**Review by itsarune — COMMENTED** [review]
can you increase the penalty_kick_enemy_play_test test timeout to 15-20 seconds so that the test ends properly, otherwise LGTM

File: src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_fsm.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+        DEFINE_SML_GUARD(setupPositionDone)
+
+        return make_transition_table(
+            // src_state + event [guard] / action = dest_state
+            *SetupPositionState_S + Update_E[!setupPositionDone_G] / setupPosition_A =
+                SetupPositionState_S,
+            SetupPositionState_S + Update_E[setupPositionDone_G] / defendKick_A =
+                DefendKickState_S,
+            DefendKickState_S + Update_E / defendKick_A, X + Update_E = X);
```
Comment (germchow) [inline src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_fsm.h:74]: Small possible nit, do we need the transition for X here? Looks like the FSM never transitions into X (unless it happens from somewhere else). Not totally sure.

Comment (williamckha) [reply src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_fsm.h:74]: I've seen it like this in other fsms in our codebase so I thought it was required to make boostsml happy -- but i could be wrong!

Comment (williamckha) [reply src/software/ai/hl/stp/play/penalty_kick_enemy/penalty_kick_enemy_play_fsm.h:74]: Tried removing the transition and looks like it's causing the CHECK in tactic.cpp to fail, so I'll leave it in for now

https://github.com/UBC-Thunderbots/Software/blob/d33056744b445243cfe7dc759677ff8e93eea984/src/software/ai/hl/stp/tactic/tactic.cpp#L42-L44

**Review by nimazareian — APPROVED** [review]
LGTM

**Review by germchow — APPROVED** [review]
LGTM

**Pull request merged**